### PR TITLE
Function boundary conditions weakly imposed in the matrix-free solver

### PR DESCRIPTION
**Pull request by blaisb**
### Description

The matrix-free solver did not support weak function boundary conditions nor outlet. This is because we did not have the infrastructure to implement these features in the past. This PR adds the function weak capability to the matrix free solver. A PR will follow in this month to add the outlet boundary condition, making the matrix free solver feature complete for single phase newtonian flow.

NOTE: This will break the Tidy CI because it still relies on the deal.II 9.5.1 backport PPA. I will live with these consequences :)

### Testing

I have added two application test to test the feature. The first one is an MMS with a single weak boundary condition imposed to zero. The second one is a taylor-couette which features mixed boundary conditions (regular no slip and function based)

### Documentation

No additional documentation are required. However, I have also added assert so that the lethe-fluid-matrix-free solver throws when an invalid boundary condition is used.

Code related list:
- [X] All in-code documentation related to this PR is up to date (Doxygen format)
- [X] Lethe documentation is up to date
- [X] New feature has unit test(s) (preferred) or application test(s), and restart files are in the generator folder
- [X] The branch is rebased onto master
- [X] Changelog (CHANGELOG.md) is up to date
- [X] Code is indented with indent-all and .prm files (examples and tests) with prm-indent

Pull request related list:
- [X] Labels are applied
- [X] There are at least 2 reviewers (or 1 if small feature) excluding the responsible for the merge
- [X] If this PR closes an issue or is related to a project, it is linked in the "Projects" or "Development" section
- [X] If the fix is temporary, an issue is opened
- [X] The PR description is cleaned and ready for merge

### CODE REVIEW

File: include/solvers/fluid_dynamics_matrix_free_operators.h [hunk 1]
```diff
@@ -77,19 +78,22 @@ class NavierStokesOperatorBase : public Subscriptor
    * jacobian on or off.
    * @param[in] enable_hessians_residual Flag to turn hessian terms from
    * residual on or off.
+   * @param[in] enable_face_terms Flag to enable face terms in the
+   * calculation of the residual and the jacobian
```
Comment (peterrum) [inline include/solvers/fluid_dynamics_matrix_free_operators.h:82]: You seem to describe the wrong input parameter.

File: include/solvers/fluid_dynamics_matrix_free_operators.h [hunk 2]
```diff
@@ -111,19 +115,23 @@ class NavierStokesOperatorBase : public Subscriptor
...
```
Comment (peterrum) [inline include/solvers/fluid_dynamics_matrix_free_operators.h:119]: Here as well.

File: include/solvers/fluid_dynamics_matrix_free_operators.h [hunk 3]
```diff
@@ -342,6 +350,58 @@ class NavierStokesOperatorBase : public Subscriptor
...
+   * @brief Loop over all internal face batches within certain range and perform a face
+   * integral with access to global vectors, i.e., gathering and scattering
+   * values. The current implementation does nothing for CG methods.
+   *
+   * @param[in] matrix_free Object that contains all data.
+   * @param[in,out] dst Global vector where the final result is added.
+   * @param[in] src Input vector with all values in all cells.
+   * @param[in] range Range of the cell batch.
+   */
+
```
Comment (peterrum) [inline include/solvers/fluid_dynamics_matrix_free_operators.h:363]:
```suggestion
```

File: include/solvers/fluid_dynamics_matrix_free_operators.h [hunk 4]
```diff
@@ -342,6 +350,58 @@ class NavierStokesOperatorBase : public Subscriptor
...
+   * values. This is used to weakly impose dirichlet boundary conditions.
+   *
+   * @tparam assemble_residual Flag to assemble the residual or the jacobian.
+   *
+   * @param[in] matrix_free Object that contains all data.
+   * @param[in,out] dst Global vector where the final result is added.
+   * @param[in] src Input vector with all values in all cells.
+   * @param[in] range Range of the cell batch.
+   */
+
```
Comment (peterrum) [inline include/solvers/fluid_dynamics_matrix_free_operators.h:383]:
```suggestion
```

File: include/solvers/fluid_dynamics_matrix_free_operators.h [hunk 5]
```diff
@@ -342,6 +350,58 @@ class NavierStokesOperatorBase : public Subscriptor
...
 
+  /**
+   * @brief Loop over all internal face batches within certain range and perform a face
+   * integral with access to global vectors, i.e., gathering and scattering
+   * values. The current implementation does nothing for CG methods.
+   *
+   * @param[in] matrix_free Object that contains all data.
+   * @param[in,out] dst Global vector where the final result is added.
+   * @param[in] src Input vector with all values in all cells.
+   * @param[in] range Range of the cell batch.
```
Comment (peterrum) [inline include/solvers/fluid_dynamics_matrix_free_operators.h:361]:
```suggestion
   * @param[in] range Range of the face batch.
```

File: include/solvers/fluid_dynamics_matrix_free_operators.h [hunk 6]
```diff
@@ -461,6 +531,14 @@ class NavierStokesOperatorBase : public Subscriptor
    */
   Table<2, Tensor<1, dim, VectorizedArray<number>>> forcing_terms;
 
+  /**
+   * @brief Flag to turn the calculation of face term on or off.
+   * This is used for weakly imposed Dirichlet boundary conditions or outlets
+   * (not implemented yet).
+   *
```
Comment (peterrum) [inline include/solvers/fluid_dynamics_matrix_free_operators.h:538]:
```suggestion
```

File: include/solvers/fluid_dynamics_matrix_free_operators.h [hunk 7]
```diff
@@ -493,6 +571,23 @@ class NavierStokesOperatorBase : public Subscriptor
   Table<2, Tensor<1, dim + 1, VectorizedArray<number>>>
     time_derivatives_previous_solutions;
 
+  /**
+   * @brief Table with correct alignment for vectorization to store the values
+   * of the previous newton step at the faces.
+   *
```
Comment (peterrum) [inline include/solvers/fluid_dynamics_matrix_free_operators.h:577]:
```suggestion
```

File: source/solvers/fluid_dynamics_matrix_free_operators.cc [hunk 8]
```diff
@@ -213,6 +244,9 @@ NavierStokesOperatorBase<dim, number>::reinit(
 
   this->enable_hessians_residual = enable_hessians_residual;
 
+  this->enable_face_terms = enable_face_terms;
+
+
```
Comment (peterrum) [inline source/solvers/fluid_dynamics_matrix_free_operators.cc:249]:
```suggestion
```

File: source/solvers/fluid_dynamics_matrix_free_operators.cc [hunk 9]
```diff
@@ -674,6 +778,11 @@ NavierStokesOperatorBase<dim, number>::
   stabilization_parameter.reinit(n_cells, integrator.n_q_points);
   stabilization_parameter_lsic.reinit(n_cells, integrator.n_q_points);
 
+  nonlinear_previous_face_values.reinit(n_boundary_faces,
+                                        face_integrator.n_q_points);
+  nonlinear_previous_face_gradient.reinit(n_boundary_faces,
+                                          face_integrator.n_q_points);
```
Comment (peterrum) [inline source/solvers/fluid_dynamics_matrix_free_operators.cc:784]: This should be optional.

File: source/solvers/fluid_dynamics_matrix_free_operators.cc [hunk 10]
```diff
@@ -830,6 +986,201 @@ NavierStokesOperatorBase<dim, number>::do_cell_integral_range(
...
+      return;
+    }
+
+
+
+  const auto face = integrator.get_current_cell_index();
+
+  //  std::cout << "Integrating over face " << face << " assemble residual "
+  //            << assemble_residual << " number of quadrature points "
+  //            << integrator.n_q_points << std::endl;
```
Comment (peterrum) [inline source/solvers/fluid_dynamics_matrix_free_operators.cc:1071]:
```suggestion
```

File: source/solvers/fluid_dynamics_matrix_free_operators.cc [hunk 11]
```diff
@@ -830,6 +986,201 @@ NavierStokesOperatorBase<dim, number>::do_cell_integral_range(
...
+
+      if (dim == 2)
+        cell_size[v] = std::sqrt(4. * cell_it->measure() / M_PI) / fe_degree;
+      else if (dim == 3)
+        cell_size[v] =
+          std::pow(6 * cell_it->measure() / M_PI, 1. / 3.) / fe_degree;
+    }
+
+  VectorizedArray<number> penalty_parameter =
+    beta / std::pow(cell_size, static_cast<number>(fe_degree + 1));
```
Comment (peterrum) [inline source/solvers/fluid_dynamics_matrix_free_operators.cc:1095]: This should be precomputed.

File: source/solvers/fluid_dynamics_matrix_free_operators.cc [hunk 12]
```diff
@@ -830,6 +986,201 @@ NavierStokesOperatorBase<dim, number>::do_cell_integral_range(
...
+
+          // std::cout << " Value is : " << value << std::endl;
+
+          Tensor<1, dim, VectorizedArray<number>> velocity_value;
+
+          velocity_value[0] = evaluate_function<dim, number>(
+            boundary_conditions.bcFunctions[boundary_index].u, point_batch);
+
+          velocity_value[1] = evaluate_function<dim, number>(
+            boundary_conditions.bcFunctions[boundary_index].v, point_batch);
```
Comment (peterrum) [inline source/solvers/fluid_dynamics_matrix_free_operators.cc:1146]: This can be also precomputed, since the information is used in multiple newton steps.

File: source/solvers/fluid_dynamics_matrix_free_operators.cc [hunk 13]
```diff
@@ -830,6 +986,201 @@ NavierStokesOperatorBase<dim, number>::do_cell_integral_range(
...
+          for (unsigned int i = 0; i < dim; ++i)
+            gradient_result[d][i] -=
+              kinematic_viscosity * value[d] * normal_vector[i];
+        }
+
+
+      // If assemble_residual, we need to also assemble the target value of the
+      // weak dirichlet boundary condition which obtained from the mu_parser
+      // function contained in the boundary_conditions object
+      if constexpr (assemble_residual)
```
Comment (peterrum) [inline source/solvers/fluid_dynamics_matrix_free_operators.cc:1131]: I would suggest to remove this path. The code below should be able to be merged with the code above if you substract from `value` the value you want to obtain.

File: source/solvers/fluid_dynamics_matrix_free_operators.cc [hunk 14]
```diff
@@ -830,6 +986,201 @@ NavierStokesOperatorBase<dim, number>::do_cell_integral_range(
...
+                      EvaluationFlags::EvaluationFlags::gradients);
+
+  for (const auto q : integrator.quadrature_point_indices())
+    {
+      typename FEFaceIntegrator::value_type    value_result    = {};
+      typename FEFaceIntegrator::gradient_type gradient_result = {};
+
+      const auto normal_vector = integrator.get_normal_vector(q);
+
+      const auto value    = integrator.get_value(q);
```
Comment (peterrum) [inline source/solvers/fluid_dynamics_matrix_free_operators.cc:1107]: Like this:
```suggestion
      auto value    = integrator.get_value(q);
      
      if constexpr (assemble_residual)
        value -= table_face[face][q];
```

File: source/solvers/fluid_dynamics_matrix_free_operators.cc [hunk 15]
```diff
@@ -1090,6 +1442,7 @@ NavierStokesStabilizedOperator<dim, number>::local_evaluate_residual(
   const std::pair<unsigned int, unsigned int> &range) const
 {
   FECellIntegrator integrator(matrix_free);
+  FEFaceIntegrator face_integrator(matrix_free, true, 0);
```
Comment (peterrum) [inline source/solvers/fluid_dynamics_matrix_free_operators.cc:1445]:
```suggestion
```

**Review by peterrum — COMMENTED** [review]
A first set of comments.

File: include/solvers/fluid_dynamics_matrix_free_operators.h (hunk 7 above)
Comment (blaisb) [reply include/solvers/fluid_dynamics_matrix_free_operators.h:577]: All of the other briefs are currently in the same way in that file, so I'll keep it uniform for now.

File: source/solvers/fluid_dynamics_matrix_free_operators.cc (hunk 13 above)
Comment (blaisb) [reply source/solvers/fluid_dynamics_matrix_free_operators.cc:1143]: Great idea, did so.

File: source/solvers/fluid_dynamics_matrix_free_operators.cc (hunk 12 above)
Comment (blaisb) [reply source/solvers/fluid_dynamics_matrix_free_operators.cc:1154]: Currently I lumped it with the stabilization evaluation because I did not want to create another indirection. Since this is pre-computed for only a few faces, I don't think this will make a big difference.

File: source/solvers/fluid_dynamics_matrix_free_operators.cc [hunk 16]
```diff
@@ -830,6 +1044,130 @@ NavierStokesOperatorBase<dim, number>::do_cell_integral_range(
...
+        phi.read_dof_values(src);
+
+      do_local_weak_dirichlet_bc<assemble_residual>(phi);
+
+      phi.distribute_local_to_global(dst);
+    }
+}
+
+
+// Add variant for face terms residual
```
Comment (lpsaavedra) [inline source/solvers/fluid_dynamics_matrix_free_operators.cc:1090]: Could you add here the equations being calculated here? I know they are in the comments below, but I find it useful to have them here as well. Then it would be uniform and similar to the functions in the `NavierStokesStabilizedOperator`

**Comment by lpsaavedra** [comment]
> NOTE: This will break the Tidy CI because it still relies on the deal.II 9.5.1 backport PPA. I will live with these consequences :)

Regarding this, can we get rid of those errors by checking the deal.II version as we do in other parts of the matrix-free application using `DEAL_II_VERSION_GTE(9, 7, 0)`?

File: applications_tests/lethe-fluid-matrix-free/cylinder_kelly_estimator.prm
[the remark this answers is not in the document]
Comment (blaisb) [reply applications_tests/lethe-fluid-matrix-free/cylinder_kelly_estimator.prm:1]: Because i removed the "false" outlet boundary condition. So the output of the simulation does not output the force on the outlet. The test was actually a bad test, because an outlet boundary condition was specified without any outlet term being applied to it. Now an assert is thrown when outlets are used.

File: applications_tests/lethe-fluid-matrix-free/cylinder_kelly_estimator.prm [hunk 17]
```diff
@@ -108,10 +108,10 @@ subsection boundary conditions
...
     set type = slip
   end
-  subsection bc 3
-    set beta = 1
-    set type = outlet
-  end
+  #subsection bc 3
+  #  set beta = 1
+  #  set type = outlet
+  #end
```
[the remark this answers is not in the document]
Comment (blaisb) [reply applications_tests/lethe-fluid-matrix-free/cylinder_kelly_estimator.prm:114]: It's needed because the matrix free method does not support outlets right now. If you don't mind I would leave it commented, I plan on adding the outlet soon and then I would revert back this test to it's original form.

File: source/solvers/fluid_dynamics_matrix_free_operators.cc [hunk 18]
```diff
@@ -830,6 +1044,130 @@ NavierStokesOperatorBase<dim, number>::do_cell_integral_range(
...
+      typename FEFaceIntegrator::value_type    value_result    = {};
+      typename FEFaceIntegrator::gradient_type gradient_result = {};
+
+      const auto normal_vector = integrator.get_normal_vector(q);
+
+      auto       value    = integrator.get_value(q);
+      const auto gradient = integrator.get_gradient(q);
+
+      // If we are assembling the residual, substract the target velocity from
+      // the velocity value This streamlines the code.
```
[the remark this answers is not in the document]
Comment (blaisb) [reply source/solvers/fluid_dynamics_matrix_free_operators.cc:1142]: Meaning that this makes the code simpler, but I can remove that line if you in find that confusing.

**Comment by blaisb** [comment]
> > NOTE: This will break the Tidy CI because it still relies on the deal.II 9.5.1 backport PPA. I will live with these consequences :)
> 
> Regarding this, can we get rid of those errors by checking the deal.II version as we do in other parts of the matrix-free application using `DEAL_II_VERSION_GTE(9, 7, 0)`?

I think so yes, let met try it out :)

File: include/solvers/fluid_dynamics_matrix_free_operators.h [hunk 19]
```diff
@@ -342,6 +351,56 @@ class NavierStokesOperatorBase : public Subscriptor
     const VectorType                            &src,
     const std::pair<unsigned int, unsigned int> &range) const;
 
+  /**
+   * @brief Loop over all internal face batches within certain range and perform a face
+   * integral with access to global vectors, i.e., gathering and scattering
+   * values. The current implementation does nothing for CG methods.
```
[the remark this answers is not in the document]
Comment (blaisb) [reply include/solvers/fluid_dynamics_matrix_free_operators.h:357]: There are no internal faces for continuous galerkin methods, so this does nothing for CG method. I will try to make it clearer.

File: source/solvers/fluid_dynamics_matrix_free_operators.cc (hunk 16 above)
Comment (blaisb) [reply source/solvers/fluid_dynamics_matrix_free_operators.cc:1089]: Sure thing :)

File: include/solvers/fluid_dynamics_matrix_free_operators.h [hunk 20]
```diff
@@ -73,23 +74,27 @@ class NavierStokesOperatorBase : public Subscriptor
    * @param[in] stabilization Stabilization type specified in parameter file.
    * @param[in] mg_level Level of the operator in case of MG methods.
    * @param[in] simulation_control Required to get the time stepping method.
+   * @param[in] boundary_conditions Contains information regarding all boundary
+   conditions. Required to weakly impose boundary conditions.
```
Comment (peterrum) [inline include/solvers/fluid_dynamics_matrix_free_operators.h:78]:
```suggestion
   * conditions. Required to weakly impose boundary conditions.
```

File: include/solvers/fluid_dynamics_matrix_free_operators.h [hunk 21]
```diff
@@ -107,23 +112,27 @@ class NavierStokesOperatorBase : public Subscriptor
    * @param[in] stabilization Stabilization type specified in parameter file.
    * @param[in] mg_level Level of the operator in case of MG methods.
    * @param[in] simulation_control Required to get the time stepping method.
+   * @param[in] boundary_conditions Contains information regarding all
+   boundary conditions. Required to weakly impose boundary conditions.
```
Comment (peterrum) [inline include/solvers/fluid_dynamics_matrix_free_operators.h:116]:
```suggestion
   * boundary conditions. Required to weakly impose boundary conditions.
```

File: source/solvers/fluid_dynamics_matrix_free_operators.cc [hunk 22]
```diff
@@ -410,8 +441,23 @@ NavierStokesOperatorBase<dim, number>::vmult(VectorType       &dst,
         .local_element(edge_constrained_indices[i]) = 0.;
     }
 
+#if DEAL_II_VERSION_GTE(9, 6, 0)
```
Comment (peterrum) [inline source/solvers/fluid_dynamics_matrix_free_operators.cc:444]: Is there somewhere an assert that weak dbc is not working for older versions of deal.II?

**Review by peterrum — APPROVED** [review]

File: source/solvers/fluid_dynamics_matrix_free_operators.cc (hunk 22 above)
Comment (blaisb) [reply source/solvers/fluid_dynamics_matrix_free_operators.cc:444]: Yeah I added an assertion elsewhere.  The code will crash now in Deal.ii 9.5 or below

**Comment by blaisb** [comment]
@lpsaavedra this is ready for your final review :)

File: source/solvers/fluid_dynamics_matrix_free_operators.cc (hunk 18 above)
[the remark this answers is not in the document]
Comment (lpsaavedra) [reply source/solvers/fluid_dynamics_matrix_free_operators.cc]: Yes, I think it is not necessary to write this here

[the remark this answers is not in the document]
Comment (blaisb) [reply source/solvers/fluid_dynamics_matrix_free_operators.cc]: Yeah I removed it

File: include/solvers/fluid_dynamics_matrix_free_operators.h (hunk 19 above)
[the remark this answers is not in the document]
Comment (lpsaavedra) [reply include/solvers/fluid_dynamics_matrix_free_operators.h:357]: I think it would be easier to understand if you just write: "Only required for compilation and not needed for our CG implementation."

File: source/solvers/fluid_dynamics_matrix_free_operators.cc [hunk 23]
```diff
@@ -830,6 +1078,137 @@ NavierStokesOperatorBase<dim, number>::do_cell_integral_range(
...
+
+      phi.distribute_local_to_global(dst);
+    }
+#endif
+}
+
+
+// Assembles weak dirichlet boundary condition using Nitche's SI method.
+// This is achieved by adding the following to the residual:
+// (v,beta (u-u_target)) - ν(v,∇δu·n) - ν(∇v·n,(u-u_target))
```
Comment (lpsaavedra) [inline source/solvers/fluid_dynamics_matrix_free_operators.cc:1129]:
```suggestion
// Assembles weak Dirichlet boundary conditions using Nitsche's SI method.
// This is achieved by adding the following to the residual:
// (v,beta (u-u_target)) - ν(v,∇δu·n) - ν(∇v·n,(u-u_target))
```

Comment (lpsaavedra) [reply source/solvers/fluid_dynamics_matrix_free_operators.cc:1129]: What is SI?

**Review by lpsaavedra — COMMENTED** [review]
Just a few comments and then it should be good to go from my side!

**Comment by blaisb** [comment]
@lpsaavedra  last comments were addressed

**Review by lpsaavedra — APPROVED** [review]
